go/tasks/plugins/webapi/agent/plugin.go

```diff
@@ -169,17 +171,38 @@ func (p Plugin) Status(ctx context.Context, taskCtx webapi.StatusContext) (phase
 	case admin.State_RETRYABLE_FAILURE:
 		return core.PhaseInfoRetryableFailure(pluginErrors.TaskFailedWithError, "failed to run the job", taskInfo), nil
 	case admin.State_SUCCEEDED:
-		if resource.Outputs != nil {
-			err := taskCtx.OutputWriter().Put(ctx, ioutils.NewInMemoryOutputReader(resource.Outputs, nil, nil))
-			if err != nil {
-				return core.PhaseInfoUndefined, err
-			}
+		err = writeOutput(ctx, taskCtx, resource)
+		if err != nil {
+			logger.Errorf(ctx, "Failed to write output with err %s", err.Error())
+			return core.PhaseInfoUndefined, err
 		}
 		return core.PhaseInfoSuccess(taskInfo), nil
 	}
 	return core.PhaseInfoUndefined, pluginErrors.Errorf(core.SystemErrorCode, "unknown execution phase [%v].", resource.State)
 }
 
+func writeOutput(ctx context.Context, taskCtx webapi.StatusContext, resource *ResourceWrapper) error {
+	taskTemplate, err := taskCtx.TaskReader().Read(ctx)
+	if err != nil {
+		return err
+	}
+
+	if taskTemplate.Interface == nil || taskTemplate.Interface.Outputs == nil || taskTemplate.Interface.Outputs.Variables == nil {
+		logger.Infof(ctx, "The task declares no outputs. Skipping writing the outputs.")
+		return nil
+	}
+
+	var opReader io.OutputReader
+	if resource.Outputs != nil {
+		logger.Infof(ctx, "Agent returned an output")
+		opReader = ioutils.NewInMemoryOutputReader(resource.Outputs, nil, nil)
+	} else {
+		logger.Infof(ctx, "Agent didn't return any output, assuming file based outputs.")
+		opReader = ioutils.NewRemoteFileOutputReader(ctx, taskCtx.DataStore(), taskCtx.OutputWriter(), taskCtx.MaxDatasetSizeBytes())
+	}
+	return taskCtx.OutputWriter().Put(ctx, opReader)
```

Comment:
can you give an example of these locations? I'm not following here - we're using the `taskCtx.OutputWriter` to construct an `OutputReader` that just writes back to the `taskCtx.OutputWriter`?

Reply:
TBH, I'm a bit confused as well. Most of our plugins do this.
- [k8s plugin](https://github.com/flyteorg/flytepropeller/blob/8446baf92748f6079f43aaa94d1b0b97df233a9e/pkg/controller/nodes/task/k8s/plugin_manager.go#L291-L297)
- [map task](https://github.com/flyteorg/flyteplugins/blob/f577f62727a16a282edfa7a49d985e47a4ccef79/go/tasks/plugins/array/outputs.go#L232-L237)
- [sagemaker](https://github.com/flyteorg/flyteplugins/blob/e7bf9a4e1b43ec7be55411fcc31babbe6c769db4/go/tasks/plugins/k8s/sagemaker/custom_training.go#L194-L199)

1. flytekit writes the proto to s3
2. propeller reads proto from s3
3. propeller writes the proto to another s3 path

However, if we don't call `OutputWriter().Put()`, the downstream task won't be able to find the input.

Reply:
error: `Last Error: UNKNOWN::Output Reader was nil. Plugin/Platform problem.`

Reply:
Update:
OutputWriter is a [NewBufferedOutputWriter](https://github.com/flyteorg/flyteplugins/blob/e7bf9a4e1b43ec7be55411fcc31babbe6c769db4/go/tasks/pluginmachinery/ioutils/buffered_output_writer.go#L9), so it won't actually write the output to another s3 path, it just [adds the output reader to NewBufferedOutputWriter](https://github.com/flyteorg/flyteplugins/blob/e7bf9a4e1b43ec7be55411fcc31babbe6c769db4/go/tasks/pluginmachinery/ioutils/buffered_output_writer.go#L15), and then propeller will validate the output by reading the data from output reader, which is in the output writer.

Reply:
thanks for figuring this out! looks great.